libbpf-tools/syscount.c

```diff
@@ -0,0 +1,1167 @@
+//
+// Based on syscount(8) from BCC by Sasha Goldshtein
+#include <sys/resource.h>
+#include <unistd.h>
+#include <signal.h>
+#include <fcntl.h>
+#include <time.h>
+#include <argp.h>
+#include <bpf/bpf.h>
+#include "syscount.h"
+#include "syscount.skel.h"
+
+#define MAX(x, y) (((x) > (y)) ? (x) : (y))
+#define warn(...) fprintf(stderr, __VA_ARGS__)
+
+const char *argp_program_version = "syscount 0.1";
+const char *argp_program_bug_address = "<bpf@vger.kernel.org>";
+static const char argp_program_doc[] =
+"\nsyscount: summarize syscall counts and latencies\n"
+"\n"
+"EXAMPLES:\n"
+"    syscount                 # print top 10 syscalls by count every second\n"
+"    syscount -p $(pidof dd)  # look only at a particular process\n"
+"    syscount -L              # measure and sort output by latency\n"
+"    syscount -P              # group statistics by pid, not by syscall\n"
+"    syscount -x -i 5         # count only failed syscalls\n"
+"    syscount -e ENOENT -i 5  # count only syscalls failed with a given errno"
+;
+
+static const struct argp_option opts[] = {
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{ "pid", 'p', "PID", 0, "Process PID to trace" },
+	{ "interval", 'i', "INTERVAL", 0, "Print summary at this interval"
+				" (seconds), 0 for infinite wait (default)" },
+	{ "duration", 'd', "DURATION", 0, "Total tracing duration (seconds)" },
+	{ "top", 'T', "TOP", 0, "Print only the top syscalls (default 10)" },
+	{ "failures", 'x', NULL, 0, "Trace only failed syscalls" },
+	{ "latency", 'L', NULL, 0, "Collect syscall latency" },
+	{ "milliseconds", 'm', NULL, 0, "Display latency in milliseconds"
+					" (default: microseconds)" },
+	{ "process", 'P', NULL, 0, "Count by process and not by syscall" },
+	{ "errno", 'e', "ERRNO", 0, "Trace only syscalls that return this error"
+				 "(numeric or EPERM, etc.)" },
+	{ "list", 'l', NULL, 0, "Print list of recognized syscalls and exit" },
+	{},
+};
+
+static struct env {
+	bool list_syscalls;
+	bool milliseconds;
+	bool failures;
+	bool verbose;
+	bool latency;
+	bool process;
+	int filter_errno;
+	int interval;
+	int duration;
+	int top;
+	pid_t pid;
+} env = {
+	.top = 10,
+};
+
+static int get_int(const char *arg, int *ret, int min, int max)
+{
+	char *end;
+	int val;
+
+	errno = 0;
+	val = strtol(arg, &end, 10);
+	if (errno) {
+		warn("strtol: %s: %s\n", arg, strerror(errno));
+		return -1;
+	} else if (end == arg || val < min || val > max) {
+		return -1;
+	}
+	if (ret)
+		*ret = val;
+	return 0;
+}
+
+static int libbpf_print_fn(enum libbpf_print_level level,
+		const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+
+	return vfprintf(stderr, format, args);
+}
+
+static int bump_memlock_rlimit(void)
+{
+	struct rlimit rlim_new = {
+		.rlim_cur	= RLIM_INFINITY,
+		.rlim_max	= RLIM_INFINITY,
+	};
+
+	return setrlimit(RLIMIT_MEMLOCK, &rlim_new);
+}
+
+#ifdef __x86_64__
+static int errno_by_name_x86_64(const char *errno_name)
```

Comment:
how about making it a bit cleaner and general by having a list of structs, like this:

```
static struct { int err_code; const char *err_name; } errs = {
    {1, "EPERM"},
    {2, "ENOENT"},
...
};
```
This seems like something that might be used by other tools eventually and will allow bi-directional lookups. code -> name and name -> code. And will avoid any unnecessary macro.

If necessary, it would also be possible to have (optionally) case-insensitive search.